libclambcc/ClamBCLogicalCompiler/ClamBCLogicalCompiler.cpp

```diff
@@ -604,13 +604,22 @@ class LogicalCompiler
     {
         Value *V         = LI.getOperand(0);
         ConstantExpr *CE = dyn_cast<ConstantExpr>(V);
-        if (!CE || CE->getOpcode() != Instruction::GetElementPtr ||
-            CE->getOperand(0) != GV || CE->getNumOperands() != 3 ||
-            !cast<ConstantInt>(CE->getOperand(1))->isZero()) {
-            printDiagnostic("Logical signature: unsupported read", &LI);
-            return false;
+        ConstantInt * CI = nullptr;
+        if (CE) {
+            if (CE->getOpcode() != Instruction::GetElementPtr ||
+                CE->getOperand(0) != GV || CE->getNumOperands() != 3 ||
+                !cast<ConstantInt>(CE->getOperand(1))->isZero()) {
+                printDiagnostic("Logical signature: unsupported read", &LI);
+                return false;
+            }
+            CI = cast<ConstantInt>(CE->getOperand(2));
+        } else {
+            /* In this case, we are directly loading the global, 
+             * instead of using a getelementptr.
+             * It is likely that this would have been changed by O3.
+             */
+            CI = ConstantInt::get(LI.getParent()->getParent()->getParent()->getContext(), APInt(64, 0));
```

Comment:
It was previously failing because there was a check for a getelementptr instruction on the load.  That is the code generated by clang with -O0 when the file is first compiled.  -O3 removes the getelementptr instruction for offset 0, since an address calculation is not necessary.